### PR TITLE
ft: ZENKO-584 failed crr metrics routes

### PR DESCRIPTION
**Pull request by philipyoo**
Add failed crr metrics

### CODE REVIEW

**Comment by philipyoo** [comment]
Version bump dropped, once this merges, need to update https://github.com/scality/backbeat/pull/349 package.json file and dropping the drop commit in that pr

File: lib/backbeat/Metrics.js [hunk 1]
```diff
@@ -181,13 +181,63 @@ class Metrics {
...
+     * @return {undefined}
+     */
+    getFailedMetrics(details, cb, data) {
+        this._getData(details, data, (err, res) => {
+            if (err && err.type) {
+                this._logger.error('error getting metric: failures', {
+                    origin: err.emthod,
+                    method: 'Metrics.getFailedMetrics',
+                });
+                return cb(err.type.customizeDescription(err.message));
```
Comment (jonathan-gramain) [inline lib/backbeat/Metrics.js:199]: Having a special error case for this looks a little bit far-fetched to me, but not critical.

File: lib/backbeat/Metrics.js [hunk 2]
```diff
@@ -181,13 +181,63 @@ class Metrics {
...
+            }
+            if (err || res.length !== details.dataPoints.length) {
+                this._logger.error('error getting metrics: failures', {
+                    method: 'Metrics.getFailedMetrics',
+                });
+                return cb(errors.InternalError);
+            }
+
+            // Find if time since start is less than EXPIRY time
+            const timeSinceStart = (Date.now() - this._internalStart) / 1000;
```
Comment (jonathan-gramain) [inline lib/backbeat/Metrics.js:209]: Later on we should put this code into a library function.

File: tests/unit/backbeat/Metrics.js [hunk 3]
```diff
@@ -63,6 +63,14 @@ describe('Metrics class', () => {
                         size: 0,
                     },
                 },
+                failures: {
+                    description: 'Number of failed replication operations ' +
+                        '(count) and bytes (size) in the last 900 seconds',
+                    results: {
+                        count: 0,
```
Comment (jonathan-gramain) [inline tests/unit/backbeat/Metrics.js:70]: Ideally we would have a test with non-zero values, though better than nothing

**Review by jonathan-gramain — APPROVED** [review]

File: lib/backbeat/Metrics.js [hunk 4]
```diff
@@ -181,13 +181,63 @@ class Metrics {
...
+     * Get failed replication stats by ops count and size in bytes
+     * @param {object} details - route details from lib/backbeat/routes.js
+     * @param {function} cb - callback(error, data)
+     * @param {array} data - optional field providing already fetched data in
+     *   order of dataPoints mentioned for each route in lib/backbeat/routes.js
+     * @return {undefined}
+     */
+    getFailedMetrics(details, cb, data) {
+        this._getData(details, data, (err, res) => {
+            if (err && err.type) {
```
Comment (rahulreddy) [inline lib/backbeat/Metrics.js:194]: This is interesting, not a blocker but it's not consistent with how we treat errors.

File: lib/backbeat/Metrics.js [hunk 5]
```diff
@@ -181,13 +181,63 @@ class Metrics {
...
+                return cb(errors.InternalError);
+            }
+
+            // Find if time since start is less than EXPIRY time
+            const timeSinceStart = (Date.now() - this._internalStart) / 1000;
+            const timeDisplay = timeSinceStart < EXPIRY ?
+                timeSinceStart : EXPIRY;
+            const numOfIntervals = Math.ceil(timeDisplay / INTERVAL);
+
+            const d = res.map(r => (
```
Comment (rahulreddy) [inline lib/backbeat/Metrics.js:214]: What does `res` look like, need more context here.

**Review by rahulreddy — APPROVED** [review]

**Pull request merged**